crates/wasmlink/runtime/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+            Entry::Empty { next } => next,
+            _ => unreachable!(),
+        };
+        return ret;
+    }
+
+    pub fn get(&self, idx: u32) -> Option<&T> {
+        match self.storage.get(idx as usize)? {
+            Entry::Full(b) => Some(b),
+            Entry::Empty { .. } => None,
+        }
+    }
+
+    pub fn get_mut(&mut self, idx: u32) -> Option<&mut T> {
+        match self.storage.get_mut(idx as usize)? {
+            Entry::Full(b) => Some(b),
+            Entry::Empty { .. } => None,
+        }
+    }
+
+    pub fn remove(&mut self, idx: u32) -> Option<T> {
+        let slot = self.storage.get_mut(idx as usize)?;
+        match mem::replace(slot, Entry::Empty { next: self.next }) {
+            Entry::Full(b) => {
+                self.next = idx as usize;
+                Some(b)
+            }
+            Entry::Empty { next } => {
+                *slot = Entry::Empty { next };
+                None
+            }
+        }
+    }
+}
+
+impl<T> Default for Slab<T> {
+    fn default() -> Slab<T> {
+        Slab {
+            storage: Vec::new(),
+            next: 0,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct IndexSlab {
+    slab: Slab<ResourceIndex>,
+}
+
+impl IndexSlab {
+    pub fn insert(&mut self, resource: ResourceIndex) -> u32 {
+        self.slab.insert(resource)
+    }
+
+    pub fn get(&self, slab_idx: u32) -> Option<ResourceIndex> {
+        self.slab.get(slab_idx).cloned()
+    }
+
+    pub fn remove(&mut self, slab_idx: u32) -> Option<ResourceIndex> {
+        self.slab.remove(slab_idx)
+    }
+}
+
+#[derive(Default)]
+pub struct ResourceSlab {
+    slab: Slab<Resource>,
+}
+
+#[derive(Debug)]
+struct Resource {
+    wasm: u32,
+    references: u32,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct ResourceIndex(u32);
+
+impl ResourceSlab {
+    pub fn insert(&mut self, wasm: u32) -> ResourceIndex {
+        ResourceIndex(self.slab.insert(Resource {
+            wasm,
+            references: 1,
+        }))
+    }
+
+    pub fn get(&self, idx: ResourceIndex) -> u32 {
+        self.slab.get(idx.0).unwrap().wasm
+    }
+
+    pub fn clone(&mut self, idx: ResourceIndex) {
+        let resource = self.slab.get_mut(idx.0).unwrap();
+        resource.references = resource.references.checked_add(1).unwrap();
+    }
+
+    pub fn drop(&mut self, idx: ResourceIndex) -> Option<u32> {
+        let resource = self.slab.get_mut(idx.0).unwrap();
+        assert!(resource.references > 0);
+        resource.references -= 1;
+        if resource.references != 0 {
+            return None;
+        }
+        let resource = self.slab.remove(idx.0).unwrap();
+        Some(resource.wasm)
+    }
+}
+
+lazy_static::lazy_static! {
+    static ref SLABS: Mutex<Vec<(IndexSlab, ResourceSlab)>> = Mutex::new(Vec::new());
+}
+
+#[no_mangle]
+pub extern "C" fn resource_insert(id: u32, res: u32) -> u32 {
+    let mut slabs = SLABS.lock().unwrap();
```

Comment:
Depending on the generated code size this may want to avoid the use of `Mutex` and `lazy_static` since ideally this is a pretty lean runtime module.

Additionally I think that we'll need to be super careful here about "panic safety" in that an aborting `unwrap()` means that this module could still be resumed via some other entry point. 

While I know it's not implemented, `Mutex` is poisoned by default on panics so the `.unwrap()` here may want to be `.unwrap_or_else` to remove the poison (and/or avoid using a mutex altogether)

Reply:
Generally I consider resuming a trap from the embedder to be a generally-unsound operation since there wasn't an unwind of wasm code such that destructors were called and thus may be in an incongruent state.

Still, I'm definitely open to removing the mutex.

Reply:
That's a good point. I'm sort of blindly assuming that this module, injected by the linker, needs to be "super hardened" in that it'll in theory run in a lot of places. That being said I think we should probably leave this as-is and revisit this later.

FWIW I don't think any poisoning will happen today, IIRC poisoning is implemented through destructors so if a panic happens this module is already "reentrant", but only in the same way that any wasm is reentrant after a trap happens (weird things may happen in the wasm)